### PR TITLE
[MODCON-61] Implement logic to pulling down a shared instance to create shadow one in the desired tenant

### PR DESCRIPTION
**Pull request by bekhzodkurbonboev**
## Purpose
Resolves: [MODCON-61](https://issues.folio.org/browse/MODCON-61) Implement logic to pulling down a shared instance to create shadow one in the desired tenant

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by bekhzodkurbonboev** [comment]
1. **Case 1:** sourceTenantId =/= centralTenantId 
image

2. **Case 2:** sourceTenantId == centralTenantId and GET 'inventory/instance' failed 
image

3. **Case 3:** sourceTenantId == centralTenantId and POST 'inventory/instance' failed 
image

4. **Case 4:** sourceTenantId == centralTenantId and all operations were successful 
image 
Following are shared real instance and shadow instance 
image
And database created records 
image

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java [hunk 1]
```diff
@@ -44,13 +57,50 @@ public SharingInstance getById(UUID consortiumId, UUID actionId) {
...
-    log.info("start:: sharingInstance with id: {}, instanceId: {}, sourceTenantId: {}, targetTenantId: {} successfully saved in db",
-      savedSharingInstance.getId(), savedSharingInstance.getInstanceId(), savedSharingInstance.getSourceTenantId(), savedSharingInstance.getTargetTenantId());
+    String centralTenantId = tenantService.getCentralTenantId();
+    String sourceTenantId = sharingInstance.getSourceTenantId();
+    String targetTenantId = sharingInstance.getTargetTenantId();
+    checkTenantsExistAndContainCentralTenantOrThrow(sourceTenantId, targetTenantId);
+
+    SharingInstanceEntity savedSharingInstance;
+    if (Objects.equals(centralTenantId, sourceTenantId)) {
+      FolioExecutionContext currentTenantContext = (FolioExecutionContext) folioExecutionContext.getInstance();
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:70]: what is the purpose of this variable? Doesn't the code below work with a plain folioExecutionContext field?

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java [hunk 2]
```diff
@@ -44,13 +57,50 @@ public SharingInstance getById(UUID consortiumId, UUID actionId) {
...
+        sharingInstance.setStatus(Status.ERROR);
+        sharingInstance.setError(String.format("Failed to get inventory instance with reason: %s", InventoryClient.getReason(ex)));
+        savedSharingInstance = sharingInstanceRepository.save(toEntity(sharingInstance));
+        log.info("start:: sharingInstance with instanceId: {}, sourceTenantId: {}, targetTenantId: {} failed with reason: {}",
+          sharingInstance.getInstanceIdentifier(), sourceTenantId, targetTenantId, sharingInstance.getError());
+        return converter.convert(savedSharingInstance, SharingInstance.class);
+      }
+
+      try (var context = new FolioExecutionContextSetter(prepareContextForTenant(targetTenantId, folioModuleMetadata, currentTenantContext))) {
+        String source = "folio".equalsIgnoreCase(inventoryInstance.get("source").asText()) ? "CONSORTIUM-FOLIO" : "CONSORTIUM-MARC";
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:85]: is it expected to have these hardcoded values here?

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java [hunk 3]
```diff
@@ -44,13 +57,50 @@ public SharingInstance getById(UUID consortiumId, UUID actionId) {
...
+
+    SharingInstanceEntity savedSharingInstance;
+    if (Objects.equals(centralTenantId, sourceTenantId)) {
+      FolioExecutionContext currentTenantContext = (FolioExecutionContext) folioExecutionContext.getInstance();
+      JsonNode inventoryInstance;
+
+      try (var context = new FolioExecutionContextSetter(prepareContextForTenant(sourceTenantId, folioModuleMetadata, currentTenantContext))) {
+        inventoryInstance = inventoryService.getById(sharingInstance.getInstanceIdentifier());
+      } catch (Exception ex) {
+        sharingInstance.setStatus(Status.ERROR);
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:76]: please add logging for the exception

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java [hunk 4]
```diff
@@ -44,13 +57,50 @@ public SharingInstance getById(UUID consortiumId, UUID actionId) {
...
+          sharingInstance.getInstanceIdentifier(), sourceTenantId, targetTenantId, sharingInstance.getError());
+        return converter.convert(savedSharingInstance, SharingInstance.class);
+      }
+
+      try (var context = new FolioExecutionContextSetter(prepareContextForTenant(targetTenantId, folioModuleMetadata, currentTenantContext))) {
+        String source = "folio".equalsIgnoreCase(inventoryInstance.get("source").asText()) ? "CONSORTIUM-FOLIO" : "CONSORTIUM-MARC";
+        var updatedInventoryInstance = ((ObjectNode) inventoryInstance).put("source", source);
+        inventoryService.saveInstance(updatedInventoryInstance);
+      } catch (Exception ex) {
+        sharingInstance.setStatus(Status.ERROR);
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:89]: please add logging for the exception

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java (hunk 2 above)
Comment (bekhzodkurbonboev) [reply src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:85]: 1. We have only two possible cases for shadow instance “CONSORTIUM-FOLIO” or “CONSORTIUM-MARC”. So currently I am setting former one, but yes it is not final yet.
2. _source_ field is mandatory in _'inventory/instance'_ https://s3.amazonaws.com/foliodocs/api/mod-inventory/r/inventory.html#inventory_instances_post:~:text=%22required%22%3A%20%5B%0A%20%20%20%20%22source%22%2C

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java (hunk 1 above)
Comment (bekhzodkurbonboev) [reply src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:70]: I try to stick general approach. Please see following files:
1. https://github.com/folio-org/mod-consortia/blob/bf65437b99771a7635bb487295492778b3217352/src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java#L112
2. https://github.com/folio-org/mod-consortia/blob/bf65437b99771a7635bb487295492778b3217352/src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java#L44
3. https://github.com/folio-org/mod-consortia/blob/bf65437b99771a7635bb487295492778b3217352/src/main/java/org/folio/consortia/service/impl/TenantServiceImpl.java#L92

Comment (tatsiana-tarhonskaya) [reply src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:70]: That's not some general approach, looks just like added one time and copypasted :)

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java [hunk 5]
```diff
@@ -44,13 +56,51 @@ public SharingInstance getById(UUID consortiumId, UUID actionId) {
...
+        var updatedInventoryInstance = ((ObjectNode) inventoryInstance).put("source", source);
+        inventoryService.saveInstance(updatedInventoryInstance);
+      } catch (Exception ex) {
+        log.warn("Error when posting instance with id: {}", sharingInstance.getInstanceIdentifier(), ex);
+        sharingInstance.setStatus(Status.ERROR);
+        sharingInstance.setError(String.format("Failed to post inventory instance with reason: %s", InventoryClient.getReason(ex)));
+        savedSharingInstance = sharingInstanceRepository.save(toEntity(sharingInstance));
+        log.info("start:: sharingInstance with instanceId: {}, sourceTenantId: {}, targetTenantId: {} failed with reason: {}",
+          sharingInstance.getInstanceIdentifier(), sourceTenantId, targetTenantId, sharingInstance.getError());
+        return converter.convert(savedSharingInstance, SharingInstance.class);
```
[the remark this answers is not in the document]
Comment (bekhzodkurbonboev) [reply src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java]: Fixed.

File: src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java [hunk 6]
```diff
@@ -26,11 +35,19 @@
 @Log4j2
 @RequiredArgsConstructor
 public class SharingInstanceServiceImpl implements SharingInstanceService {
+  private static final String INSTANCE_SOURCE_VALUE = "folio";
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:38]: Suggest to rename to FOLIO_SOURCE_VALUE

**Review by SerhiiNosko — APPROVED** [review]

**Review by tatsiana-tarhonskaya — APPROVED** [review]

**Pull request merged**